Sources/Algorithms/LazySplitCollection.swift

```diff
@@ -0,0 +1,367 @@
+      // Only use the lower bound to test for ordering, as above.
+      lhs.baseRange.lowerBound < rhs.baseRange.lowerBound
+    }
+  }
+
+  /// Returns the index of the subsequence starting at or after the given base collection index.
+  internal func indexForSubsequence(
+    atOrAfter lowerBound: Base.Index,
+    sequenceLength: Int = 0,
+    separatorCount: Int = 0
+  ) -> Index {
+    var newSeparatorCount = separatorCount
+    var start = lowerBound
+    // If we don't have any more splits to do (which we'll determine shortly),
+    // the end of the next subsequence will be the end of the base collection.
+    var end = base.endIndex
+
+    // The number of separators encountered thus far is identical to the number
+    // of splits performed thus far.
+    if newSeparatorCount < maxSplits {
+      // The non-inclusive end of the next subsequence is marked by the next
+      // separator, or the end of the base collection.
+      end =
+        base[start...].firstIndex(where: isSeparator)
+        ?? base.endIndex
+
+      if base[start..<end].isEmpty {
+        if omittingEmptySubsequences {
+          // Find the next subsequence of non-separators.
+          start =
+            base[end...].firstIndex(where: { !isSeparator($0) })
+            ?? base.endIndex
+          if start == base.endIndex {
+            // No non-separators left in the base collection. We're done.
+            return endIndex
+          }
+          end = base[start...].firstIndex(where: isSeparator) ?? base.endIndex
+        }
+      }
+    }
+
+    if end < base.endIndex {
+      newSeparatorCount += 1
+    }
+
+    return Index(
+      baseRange: start..<end,
+      sequenceLength: sequenceLength + 1,
+      separatorCount: newSeparatorCount
+    )
+  }
+
+  public var startIndex: Index {
+    _startIndex
+  }
+
+  public var endIndex: Index {
+    Index(baseRange: base.endIndex..<base.endIndex)
+  }
+
+  public func index(after i: Index) -> Index {
+    precondition(i != endIndex, "Can't advance past endIndex")
+
+    var subsequenceStart = i.baseRange.upperBound
+    if subsequenceStart < base.endIndex {
+      // If we're not already at the end of the base collection, the previous
+      // susequence ended with a separator. Start searching for the next
+      // subsequence at the following element.
+      subsequenceStart = base.index(after: i.baseRange.upperBound)
+    }
+
+    guard subsequenceStart != base.endIndex else {
+      if !omittingEmptySubsequences
+        && i.sequenceLength < i.separatorCount + 1
+      {
+        /// The base collection ended with a separator, so we need to emit one
+        /// more empty subsequence. Its range can't be equal to that of
+        /// `endIndex`, else we'll terminate iteration prematurely.
+        return Index(
+          baseRange: i.baseRange.upperBound..<i.baseRange.upperBound,
+          sequenceLength: i.sequenceLength + 1,
+          separatorCount: i.separatorCount
```

Comment:
Something feels off here — I think we want the final subsequence to represent the space between the last element and the end of the collection, which has the range `base.endIndex..<base.endIndex`. In addition, if we do it this way I think the `<` and `==` implementations aren’t going to be correct for this penultimate index.

Here’s an example with the existing split. If we don’t omit the empty subsequences and have a separator-only collection, the resulting ranges are all the empty gaps between elements—including `endIndex..<endIndex`:

```swift
let a = [0, 0, 0]
    .split(separator: 0, omittingEmptySubsequences: false)
    .map(\.indices)
// a == [0..<0, 1..<1, 2..<2, 3..<3]
```

Given this, we can’t depend on there being any range of `base` that would be uniquely the `endIndex` of `LazySplitCollection`. Maybe the subsequence counter == `maxSplits` or `Int.max` is how `endIndex` is designated? Otherwise, we’ve used an enumeration with an `atEnd` case in various other places. What do you think?

Reply:
I figured there would be questions about this, because I have them too! It _is_ wonky right now. I meant to put a note here, but got distracted thinking about `BidirectionalCollection` conformance.

I thought perhaps `base.endIndex..<base.endIndex` was a convention for these range-based indexes that I should maintain. But that does result in a logically duplicate index in the case where the base collection ends with at least two separators.

I pondered including `sequenceLength` in the equality comparison, which would fix that particular problem. But even when the collection ends with only one separator, and the final empty sequence is not strictly a duplicate, its range doesn't feel right--it's the empty range right before the last element, instead of after it.

```swift
[7, 42].split(separator: 42, omittingEmptySubsequences: false)
  .map(\.indices)
// -> [0..<1, 2..<2]
[7, 42].lazy.split(separator: 42, omittingEmptySubsequences: false)
  .map(\.indices)
// -> [0..<1, 1..<1]
```

So I agree that it feels right to leave `base.endIndex..<base.endIndex` available for that final empty subsequence. I'm leaning towards using an enumeration, because I think there are some problematic cases trying to use something like `sequenceLength == maxSplits + 1` or `(sequenceLength - 1) == maxSplits` (to avoid a problem at `maxSplits == Int.max`) as the determining factor.

For example, the following would work:

```swift
let parts = [7, 42].lazy.split(separator: 42, omittingEmptySubsequences: false)
// maxSplits == Int.max
// endIndex == Index(baseRange: Range(2..<2), sequenceLength: Int.max, separatorCount: ?)
parts.index(after: parts.startIndex)
// -> Index(baseRange: Range(2..<2), sequenceLength: 2, separatorCount: 1)
```

But

```swift
let parts = [7, 42].lazy.split(separator: 42, maxSplits: 1, omittingEmptySubsequences: false)
// endIndex == Index(baseRange: Range(2..<2), sequenceLength: Int.max, separatorCount: ?)
parts.index(after: parts.startIndex)
// -> Index(baseRange: Range(2..<2), sequenceLength: 2, separatorCount: 1)
```

doesn't, without stretching the meaning of one of the variables available.

Reply:
> Maybe the subsequence counter == `maxSplits` or `Int.max` is how `endIndex` is designated?

Sorry, I think I misunderstood you. When I wrote the above, I thought you were saying an index with `subsequenceLength == maxSplits`, whatever that was, would be the end index. But now I think you were saying `subsequenceLength == Int.max` is the end, which wouldn't be problematic except in a very extreme edge case. 😄

I still prefer the idea of using an enum as it would be more explicit, but maybe I'll try this first, just to see how it feels, because it's simpler.

Reply:
Defining `endIndex` as
```swift
Index(
  baseRange: base.endIndex..<base.endIndex,
  sequenceLength: Int.max,
  separatorCount: Int.max
)
```

and redefining the equality and comparison operators as
```swift
public static func == (lhs: Index, rhs: Index) -> Bool {
  // `sequenceLength` is equivalent to its index's 1-based position in the
  // collection of indices.
  lhs.sequenceLength == rhs.sequenceLength
}

public static func < (lhs: Index, rhs: Index) -> Bool {
  lhs.sequenceLength < rhs.sequenceLength
}
```

lets us achieve this goal
```swift
[42, 42, 42].lazy.split(
  separator: 42,
  omittingEmptySubsequences: false
).map(\.indices)
// -> [0..<0, 1..<1, 2..<2, 3..<3]
```

and keeps all the tests green except for one that I added out of curiosity:
```swift
func testIntsEmptyEquatableCollectionNotOmittingEmpty() {
  let empty: [Int] = []
  let expectedResult = empty.split(
    separator: 42,
    omittingEmptySubsequences: false
  )
  let testResult = empty.lazy.split(
    separator: 42,
    omittingEmptySubsequences: false
  )
  XCTAssertEqualSequences(expectedResult, testResult)
}
```
The eager split returns one empty sequence with indices `0..<0`, whereas the lazy split returns none because I set `startIndex` to `endIndex` for empty base collections.

Not sure what I think about the eager split's behavior here. It strictly obeys the rule that the number of subsequences resulting from splitting is one more than the number of separators, but there's also no separator on which to split in the first place. I'll check the implementation of the eager split, but I'm curious what people think about this.

Reply:
> The eager split returns one empty sequence with indices 0..<0, whereas the lazy split returns none because I set startIndex to endIndex for empty base collections.

😮 If I think about it, that makes sense, since splitting something with just one separator yields two empty subsequences. But it's still really odd!

Reply:
I deleted my comment about that, because it eventually occurred to me that the one empty subsequence in that case is just the original empty collection, unsplit, which is returned when not omitting empty subsequences. That makes sense to me now, but yeah, it initially seemed odd.

Reply:
According to the tests, I still have some issues to work out with the empty collection case. Then I do want to make an attempt at an enum version, to see if I prefer that.

I actually played around with an enum implementation before I submitted this PR, but abandoned it because `switch` statements didn't seem like a good fit in this algorithm, where in the majority of code paths you're not dealing with the `.atEnd` case. I really like how the enum works in `Intersperse`, but that type has a lovely symmetry where every method can be just one `switch` statement. That didn't seem to be the case here, but I want to give it another try.

Reply:
I ended up pushing the commit that redefined `endIndex` to have `sequenceLength == Int.max`. It was a fairly small change, at least in terms of diffs, that fixed a bug in the previous implementation in addition to allowing the more natural representation of a final empty subsequence, so it seemed like a win.

I'll investigate an enum-based implementation tomorrow.

Reply:
I've been playing around with an `enum`-based implementation, and I don't think it's headed toward something I want to submit. However, I don't have much experience designing types based on Swift's super-cool, super-powerful `enum`, so it might be _me_ that headed in the wrong direction to begin with.

This is the `Index` I wrote:

```swift
extension LazySplitCollection: LazyCollectionProtocol {
  /// Position of a subsequence in a split collection.
  public struct Index: Comparable {
    internal enum Representation: Equatable {
      case includingEmptySubsequences(
            /// The range corresponding to the subsequence at this position.
            baseRange: Range<Base.Index>,
            /// The number of subsequences up to and including this position in the
            /// collection.
            sequenceLength: Int,
            /// The number of separators encountered up to and inclluding this
            /// position in the collection.
            separatorCount: Int
           )
      case omittingEmptySubsequences(
            /// The range corresponding to the subsequence at this position.
            baseRange: Range<Base.Index>
           )
      case end
    }

    internal let representation: Representation
// [etc.]
```

and I like how that feels--how the associated types are exactly the properties needed for each case. But in adapting the rest of the implementation to this, I'm finding the amount of code and code duplication is growing, and I'm ending up with too many situations like

```swift
switch index {
case .omittingEmptySubsequences:
  // some code
case .includingEmptySubsequences:
  // some mostly similar code with a few crucial differences
case .end:
  // there's a precondition to ensure we never get here!
}
```

In terms of what methods the cases of this enumeration need to participate in, and the implementation for each case, the first two overlap too much, while the third barely overlaps with the first two at all. Is there a rule of thumb to be gleaned here?

But again, I don't have much experience in this area. Did I get off on the wrong foot? I can show additional, more specific code examples if my intent isn't clear.

Thank you for your patience with this PR, and I appreciate the opportunity to learn.

Reply:
I think the distinction between the omitting and not omitting case isn't as sharp as I first thought. I'm going to try just an element and end case, and see how much of the redundancy and awkwardness that eliminates.

Reply:
Sounds good — I think I would go that route if I were using an enum as well. I think in this case, since the `.max` value works well to indicate the `endIndex`, I'd just stick with that.